presidio-analyzer/presidio_analyzer/predefined_recognizers/fi_personal_identity_code_recognizer.py

```diff
@@ -0,0 +1,63 @@
+from datetime import datetime
+from typing import List, Optional
+
+from presidio_analyzer import Pattern, PatternRecognizer
+
+
+class FiPersonalIdentityCodeRecognizer(PatternRecognizer):
+    """
+    Recognizes and validates Finnish Personal Identity Codes (Henkilötunnus).
+
+    :param patterns: List of patterns to be used by this recognizer
+    :param context: List of context words to increase confidence in detection
+    :param supported_language: Language this recognizer supports
+    :param supported_entity: The entity this recognizer can detect
+    """
+
+    PATTERNS = [
+        Pattern(
+            "Finnish Personal Identity Code (Medium)",
+            r"\b(\d{6})([+-ABCDEFYXWVU])(\d{3})([0123456789ABCDEFHJKLMNPRSTUVWXY])\b",
+            0.5,
+        ),
+        Pattern(
+            "Finnish Personal Identity Code (Very Weak)",
+            r"(\d{6})([+-ABCDEFYXWVU])(\d{3})([0123456789ABCDEFHJKLMNPRSTUVWXY])",
+            0.1,
+        ),
+    ]
+    CONTEXT = ["hetu", "henkilötunnus", "personbeteckningen", "personal identity code"]
+
+    def __init__(
+        self,
+        patterns: Optional[List[Pattern]] = None,
+        context: Optional[List[str]] = None,
+        supported_language: str = "fi",
+        supported_entity: str = "FI_PERSONAL_IDENTITY_CODE",
+    ):
+        patterns = patterns if patterns else self.PATTERNS
+        context = context if context else self.CONTEXT
+        super().__init__(
+            supported_entity=supported_entity,
+            patterns=patterns,
+            context=context,
+            supported_language=supported_language,
+        )
+
+    def validate_result(self, pattern_text: str) -> Optional[bool]:
+        """Validate the pattern by using the control character."""
+
+        # More information on the validation logic from:
+        # https://dvv.fi/en/personal-identity-code
+        # Under "How is the control character for a personal identity code calculated?".
+        date_part = pattern_text[0:6]
```

Comment:
Please check for len prior to running this. If in the future the pattern changes and we pass shorter strings, it will raise an exception and stop the whole process.

Reply:
Thanks a lot for your feedback @omri374 . In my understanding, the regular expression is already taking care of that?

My understanding is that, when entering `validate_result`, we have already matched `(\d{6})([+-ABCDEFYXWVU])(\d{3})([0123456789ABCDEFHJKLMNPRSTUVWXY])` and therefore, we can be sure that the string we mainpulate (`pattern_text`) is already following the right format (`pattern_text[0:6]` are all digits). 

In other words, I understand that the regular expression is already checking the length for us? Is my understanding correct or did I misunderstand your previous comment?

Reply:
Yes, you are correct. I wasn't clear enough on my comment. Regex patterns sometimes change, and adding another check for len here is useful as we might unexpectedly break something if the pattern changes but the checksum doesn't. An exception here would not only affect the users of this recognizer, but all users of the tool. Hope this makes sense.

Reply:
Thanks @omri374. Would something like this suffice?

```
if len(pattern_text) != 11:
    return False
```

Then, the, `datetime.strptime(date_part, "%d%m%y")` will take care of validating the `date_part`.

While I understand that we would like to avoid exceptions in this context, I must admit that I fail to think of scenarios where we would change the pattern without changing the checksum and, the tests would still pass?

Reply:
Thanks a lot for your reply.

I have now added the length check.